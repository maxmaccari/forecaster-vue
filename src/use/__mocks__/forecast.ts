export const useForecast = jest.fn()