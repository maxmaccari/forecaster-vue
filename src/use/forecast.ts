declare interface WindResponse {
  speed: number
  deg: number
}

declare interface WeatherResponse {
  coord: { lon: number; lat: number }
  weather: {
    id: number
    main: string
    description: string
    icon: string
  }[]
  base: string
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    humidity: number
  }
  visibility: number
  wind: WindResponse
  clouds: {
    all: number
  }
  dt: number
  sys: {
    type: number
    id: number
    country: string
    sunrise: number
    sunset: number
  }
  timezone: number
  id: number
  name: string
  cod: number
}

declare interface ListResponse {
  dt: number
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    sea_level: number
    grnd_level: number
    humidity: number
    temp_kf: number
  }
  weather: [
    {
      id: number
      main: string
      description: string
      icon: string
    }
  ]
  clouds: { all: number }
  wind: WindResponse
  visibility: number
  pop: number
  sys: { pod: string }
  dt_txt: string
}

declare interface ForecastResponse {
  cod: string
  message: number
  cnt: number
  list: ListResponse[]
  city: {
    id: number
    name: string
    coord: { lat: number; lon: number }
    country: string
    population: number
    timezone: number
    sunrise: number
    sunset: number
  }
}

export class Weather {
  description: string | null = null
  icon: string | null = null
  temperature = 0
  minTemperature = 0
  maxTemperature = 0
  feelsLike = 0
  humidity = 0
  pressure = 0

  constructor(weather: WeatherResponse | ListResponse | null = null) {
    if (!weather) return

    this.description = weather.weather[0]
      ? weather.weather[0].description
      : null
    this.icon = weather.weather[0] ? weather.weather[0].icon : null
    this.temperature = weather.main.temp
    this.minTemperature = weather.main.temp_min
    this.maxTemperature = weather.main.temp_max
    this.feelsLike = weather.main.feels_like
    this.humidity = weather.main.humidity
    this.pressure = weather.main.pressure
  }
}

export class Wind {
  direction = 0
  speed = 0

  constructor(windResponse: WindResponse | null = null) {
    this.direction = windResponse?.deg || 0
    this.speed = windResponse?.speed || 0
  }
}

export class ForecastDetails {
  clouds = 0
  timestamp = 0
  weather: Weather = new Weather()
  wind: Wind = new Wind()

  constructor(details: ListResponse) {
    this.clouds = details.clouds.all
    this.timestamp = details.dt
    this.weather = new Weather(details)
  }
}

export class Forecast {
  name = ''
  country = ''
  weather: Weather = new Weather()
  wind: Wind = new Wind()
  details: ForecastDetails[] = []

  constructor(weather: WeatherResponse, forecast: ForecastResponse) {
    this.name = weather.name
    this.country = forecast.city.country
    this.weather = new Weather(weather)
    this.wind = new Wind(weather.wind)
    this.details = forecast.list.map(details => {
      return new ForecastDetails(details)
    })
  }
}

class CacheHash {
  [Identifier: string]: Forecast
}

let cache: CacheHash = new CacheHash()

const API_URL = process.env.VUE_APP_API_URL
const API_KEY = process.env.VUE_APP_API_KEY

export const fetchWeather = (location: string): Promise<WeatherResponse> => {
  const url = `${API_URL}/weather?q=${location}&appid=${API_KEY}&units=metric`

  return window.fetch(url).then(response => {
    return (response.json() as Record<string, any>) as WeatherResponse
  })
}

export const fetchForecast = (location: string): Promise<ForecastResponse> => {
  const url = `${API_URL}/forecast?q=${location}&appid=${API_KEY}&units=metric`

  return window.fetch(url).then(response => {
    return (response.json() as Record<string, any>) as ForecastResponse
  })
}

export const clearCache = () => {
  cache = new CacheHash()
}

export const useForecast = async (location: string): Promise<Forecast> => {
  if (cache[location]) return cache[location]

  const [weatherJson, forecastJson] = await Promise.all([
    fetchWeather(location),
    fetchForecast(location),
  ])

  const forecast = new Forecast(weatherJson, forecastJson)
  cache[location] = forecast

  return forecast
}
