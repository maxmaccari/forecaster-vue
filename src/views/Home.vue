<template>
  <div class="flex items-center justify-center h-full home">
    <div class="w-full max-w-md px-2 py-3 sm:p-6 bg-gray-translucid">
      <img class="m-auto" src="@/assets/img/logo.svg" alt="Forecaster" />

      <hr class="w-full mt-4 border-gray-500 sm:mt-5" />

      <form
        @submit.prevent="goToForecast"
        class="flex w-full mt-4 sm:mt-5"
        data-test-id="locationForm"
      >
        <input
          type="text"
          class="flex-1 input"
          placeholder="Where are you now?"
          data-test-id="locationInput"
          v-model.trim="location"
        />
        <button
          type="submit"
          class="ml-2 btn"
          :disabled="isGoDisabled"
          data-test-id="goButton"
        >
          Go
        </button>
      </form>

      <hr class="w-full mt-4 border-gray-500 sm:mt-5" />
      <p class="mt-3 font-thin text-blue-900 sm:mt-4">
        Designed and developed by
        <a
          href="https://www.linkedin.com/in/maxmaccari/"
          class="link"
          target="_blank"
          >Maxsuel Maccari</a
        >. You can check all credits of used assets
        <router-link :to="{ name: 'Credits' }" class="link">here</router-link>.
      </p>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

const Component = defineComponent({
  name: 'Home',
  components: {},
  data() {
    return {
      location: '',
    }
  },
  computed: {
    isGoDisabled(): boolean {
      return !this.location
    },
  },
  methods: {
    goToForecast() {
      this.$router.push({
        name: 'Forecast',
        params: { location: this.location.toLowerCase() },
      })
    },
  },
})

export default Component
</script>
