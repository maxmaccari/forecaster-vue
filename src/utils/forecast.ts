import { Forecast } from '@/use/forecast'

export class NextHoursSummary {
  constructor(
    public temperature: number,
    public time: string,
    public clouds: number,
    public icon: string | null
  ) {}
}

export class NextWeekEntry {
  constructor(
    public day: string,
    public clouds: number,
    public min: number,
    public temperature: number,
    public icon: string | null,
    public description: string | null
  ) {}
}

export const dateFormat = new Intl.DateTimeFormat('en-US', {
  day: 'numeric',
  weekday: 'long',
})

export const timeFormat = new Intl.DateTimeFormat('en-US', {
  hour: 'numeric',
  minute: 'numeric',
})

export const getNextHours = (forecast: Forecast): NextHoursSummary[] => {
  return [forecast.details[0], forecast.details[2], forecast.details[4]].map(
    detail => new NextHoursSummary(
      Math.round(detail.weather.temperature),
      timeFormat.format(new Date(detail.timestamp * 1000)),
      detail.clouds,
      detail.weather.icon,
    )
  )
}

export const getNextWeek = (forecast: Forecast): NextWeekEntry[] => {
  return forecast.details.reduce((week: NextWeekEntry[], detail) => {
    const date = new Date(detail.timestamp * 1000)
    const day = dateFormat.format(date)
    const forecast = week.find(forecast => forecast.day === day)

    if (!forecast) {
      week.push(new NextWeekEntry(day, 
        detail.clouds,
        Math.round(detail.weather.minTemperature),
        Math.round(detail.weather.temperature),
        detail.weather.icon,
        detail.weather.description,
      ))
    } else {
      if (detail.weather.minTemperature < forecast.min) {
        forecast.min = Math.round(detail.weather.minTemperature)
      }

      if (detail.weather.temperature > forecast.temperature) {
        forecast.temperature = Math.round(detail.weather.temperature)
      }

      if (date.getHours() > 10 && date.getHours() < 14) {
        forecast.icon = detail.weather.icon
      }
    }

    return week
  }, [])
}
