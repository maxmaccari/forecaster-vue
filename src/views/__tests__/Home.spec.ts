import { describe, it, expect, vi, beforeEach } from 'vitest';

import { shallowMount } from '@vue/test-utils'
import Home from '../Home.vue'
import { clearCache } from '@/use/forecast'

vi.mock('@/use/forecast')

const router = {
  push: vi.fn(),
}

const createWrapper = (options = {}) => {
  return shallowMount(Home, {
    global: {
      mocks: { $router: router },
      components: {
        RouterLink: {},
        AppLogo: {},
      },
    },
    ...options,
  })
}

describe('Home', () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  it('renders the view properly', () => {
    const wrapper = createWrapper()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('disables go button if location is empty', () => {
    const wrapper = createWrapper()

    const goButton = wrapper.find('[data-test-id="goButton"]')
    expect(goButton.element.hasAttribute('disabled')).toBe(true)
  })

  it('enables go button if location is filled', async () => {
    const wrapper = createWrapper()

    const locationInput = wrapper.find('[data-test-id="locationInput"')
    await locationInput.setValue('New York')

    const goButton = wrapper.find('[data-test-id="goButton"]')
    expect(goButton.element.hasAttribute('disabled')).toBe(false)
  })

  it('pushes to Forecast route if go button is pressed with inputted location', async () => {
    const wrapper = createWrapper()

    const locationInput = wrapper.find('[data-test-id="locationInput"')
    await locationInput.setValue('  New York ')

    const locationForm = wrapper.find('[data-test-id="locationForm"]')
    await locationForm.trigger('submit')

    expect(router.push).toBeCalledWith({
      name: 'Forecast',
      params: { location: 'new york' },
    })
  })

  it('clears the forecast cache when it is called', async () => {
    const wrapper = createWrapper()

    expect(clearCache).toBeCalledTimes(1)
  })
})
