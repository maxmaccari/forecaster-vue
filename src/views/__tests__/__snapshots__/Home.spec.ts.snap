// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > renders the view properly 1`] = `
"<div class="flex items-center justify-center h-full home">
  <div class="w-full max-w-md px-2 py-3 mx-4 sm:mx-0 sm:p-6 bg-secondary/[0.8]">
    <app-logo-stub class="m-auto"></app-logo-stub>
    <hr class="w-full mt-4 border-gray sm:mt-5">
    <form class="flex w-full mt-4 sm:mt-5" data-test-id="locationForm"><input type="text" class="flex-1 input" placeholder="Where are you now?" data-test-id="locationInput"><button type="submit" class="ml-2 btn" disabled="" data-test-id="goButton"> Go </button></form>
    <!--v-if-->
    <hr class="w-full mt-4 border-gray sm:mt-5">
    <p class="mt-3 text-base leading-6 font-light text-primary sm:mt-4 text-justify"> Designed and developed by <a href="https://www.linkedin.com/in/maxmaccari/" class="link" target="_blank">Maxsuel Maccari</a>. You can check all credits of used assets <router-link-stub to="[object Object]" class="link"></router-link-stub>. </p>
  </div>
</div>"
`;
